elasticapm/contrib/starlette/utils.py

```diff
@@ -0,0 +1,162 @@
+
+import json
+
+from starlette.requests import Request
+from starlette.types import Message
+from starlette.responses import Response
+
+from elasticapm.conf import constants
+from elasticapm.utils import compat, get_url_dict
+
+
+async def get_data_from_request(request: Request, capture_body=False, capture_headers=True) -> dict:
+    """Loads data from incoming request for APM capturing.
+
+    Args:
+        request (Request)
+        capture_body (bool): Loads also request body if True
+        capture_headers (bool): Loads also request headers if True
+
+    Returns:
+        dict
+    """
+    result = {
+        "method": request.method,
+        "socket": {
+            "remote_address": _get_client_ip(request),
+            "encrypted": request.url.is_secure
+        },
+        "cookies": request.cookies,
+    }
+    if capture_headers:
+        result["headers"] = dict(request.headers)
+
+    if request.method in constants.HTTP_WITH_BODY:
+        if request.headers['content-type'] == "application/x-www-form-urlencoded":
+            query_params = await get_body(request)
+            body = await query_params_to_dict(query_params)
+        else:
+            body = await get_body(request)
+            body = json.loads(body)
+        if body is not None:
+            result["body"] = body if capture_body else "[REDACTED]"
+
+    result["url"] = get_url_dict(str(request.url))
+
+    return result
+
+
+async def get_data_from_response(response: Response, capture_body=False, capture_headers=True) -> dict:
+    """Loads data from response for APM capturing.
+
+    Args:
+        response (Response)
+        capture_body (bool): Loads also response body if True
+        capture_headers (bool): Loads also response HTTP headers if True
+
+    Returns:
+        dict
+    """
+    result = {}
+
+    if isinstance(getattr(response, "status_code", None), compat.integer_types):
+        result["status_code"] = response.status_code
+
+    if capture_headers and getattr(response, "headers", None):
+        headers = response.headers
+        result["headers"] = {key: ";".join(headers.getlist(key)) for key in compat.iterkeys(headers)}
+
+    if capture_body:
+        result["body"] = response.body.decode("utf-8")
+
+    return result
+
+
+async def set_body(request: Request, body: bytes):
+    """Overwrites body in Starlette.
+
+    Args:
+        request (Request)
+        body (bytes)
+    """
+    async def receive() -> Message:
+        return {"type": "http.request", "body": body}
+
+    request._receive = receive
+
+
+async def get_body(request: Request) -> str:
+    """Gets body from the request.
+
+    todo: This is not very pretty however it is not usual to get request body out of the target method (business logic).
+
+    Args:
+        request (Request)
+
+    Returns:
+        str
+    """
+    body = await request.body()
+    await set_body(request, body)
+
+    return body.decode("utf-8")
```

Comment:
I don't think this is quite right.

The problem with this solution is that when you call `request.body()`, that function [uses `request.stream()` to get the body](https://github.com/encode/starlette/blob/20973f2ed31f4fe7faca988f8590fe3eaf70541e/starlette/requests.py#L194-L195). Because you haven't reset the state around streams, if the user then calls `request.stream()` they will get an error.

I think if you set `request._stream_consumed = False` that will make this work.

Alternatively, you could just set `request._body` and not touch `_receive` at all. As far as I can tell, everything will work, as `request._body` [short-circuits streaming](https://github.com/encode/starlette/blob/20973f2ed31f4fe7faca988f8590fe3eaf70541e/starlette/requests.py#L169-L172). That removes the need for `set_body` entirely, unless I'm missing something.

Reply:
I admit I hadn't tested as much as I got that problem.
Loading of request body out of the method of business logic (e.g. a middleware) seems an antipattern to me. We shouldn't do it in async world at all. However, this is great information for APM tracking, especially in POST methods.

Thanks for advice I'll try the setting of `request._stream_consumed = False`.

Reply:
What do you think about my alternate solution, just setting `request._body` with the body? That way you could avoid overriding `_receive` completely. Seems cleaner to me, but I want to make sure I'm not missing some reason that's not the best solution.

Reply:
Honestly, I think both of the solutions are wrong and it doesn't matter which one.
however, you're the owner of this repository ;-) and if you want me to change it I'll do it...

Reply:
Updated..

Reply:
@basepi I've been testing the middleware in my app and the last solution of set `request._body` doesn't work.  The object of type `starlete.requests.Request` is instanced more times in Starlete (I didn't look for where). If I invoke the request to my app the Starlete framework waits to load (stream) body but it's not able to. 
I'll revert it to the first solution ;-)

Reply:
Heh, it doesn't work due to the sequence of operations in async probably.

Reply:
Fascinating. Thank you for testing my solution, I'm ok with the first solution.

We'll see if any other users run into issues with this. We may have to put it behind a configuration gate if they do. It's definitely not pretty, but like you said, it's "wrong" anyway, but useful.

Reply:
Fascinating. Thank you for testing my solution, I'm ok with the first solution.

We'll see if any other users run into issues with this. We may have to put it behind a configuration gate if they do. It's definitely not pretty, but like you said, it's "wrong" anyway, but useful.